airflow/models/dag.py

```diff
@@ -1373,6 +1374,40 @@ def normalized_schedule_interval(self) -> ScheduleInterval:
             _schedule_interval = self.schedule_interval
         return _schedule_interval
 
+    @staticmethod
+    @internal_api_call
+    @provide_session
+    def fetch_callback(
+        dag: DAG,
+        dagrun: DagRun | DagRunPydantic,
```

Comment:
I think this is the place where we have a little trouble, but the solution is easy.  We are not supposed to send `DagRun | DagRunPydantic` classes from client to server. What we are supposed to do is to replace it with "dag_run_id: str" and then locally get the dagrun as usual from SQLalchemy session.

This is very efficient when it comes to "db" mode. If the object is already present in the session, it should just be returned without any DB operations involved (so we will have the same DagRun object that the caller has). In case of the DB-mode, it will be, on-the-other hand retrieved from the DB as expected.

The only thing we have to make sure about (and this is something when we verify when we run tests in db-less mode) is to make sure that this DagRun object in the caller has not been modified since it was retrieved - if it was then such a modification (in DagFileProcessor or Triggerer) should be replaced with an internal_api method call to do such a modification. I think those cases will be extremely rare though (if at all). Most of such modifications happens in Scheduler, not DagFileProcessor and Triggerer. Only LocalTaskRunner does effectively modify DagRun and those are things that we will have to carefully wrap in an internal_api_call method.

I think this is something that we should realise (and maybe even enforce) - while we can return DagRun | DagRunPydantic (and friends) from `internal_api` decorated method - we should never pass then  as an argument. We should only pass "primitive" values like strings and regular objects (like DAG) - but not potentially ORM models.

Reply:
This will also help to clarify that in all the methods only called from the `internal_api` decorated methods, we should just bass the ORM objects. There were a few methods below where we missed typing for taskinstance and dagrun - and they are just called from "inside" of the `internal_api` calls - and they should simply get a TaskInstance and DagRun as parameters.

It really only makes sense to "retrieve" the DagRunPydantic objects to client, but any updates we make we should make as `update(dag_run_id: str, with_this_information: ....)` kind of method. While Pydantic provides us with an easy way to convert from ORM objects to Pydantic ones, the conversion back is not easy and we do not even really need to do it.

cc: @uranusjr and  @mhenc also for discussion on it.